syntax = "proto3";

import "google/protobuf/empty.proto";

option java_multiple_files = true;
option java_package = "io.grpc.snippet_sample";
option java_outer_classname = "snippet_sample_Proto";
option objc_class_prefix = "SSP";

package snippetsample;

service SnippetSample {
  rpc SetSnippet (stream SnippetRequest) returns (stream Result) {}
  rpc Run (Request) returns (Result) {}
}

message SnippetRequest {
  int32 type = 1;
  Snippet snippet = 2;
}

message Snippet {
  enum ValueType {
    INT8 = 0;
    INT16 = 1;
    INT32 = 2; //DEFAULT
    INT64 = 3;
    FLOAT32 = 4;
    FLOAT64 = 5;
    NUMERIC = 6;
    DATE = 7;
    TIMESTAMP = 8;
    STRING = 9;
    COLUMN = 10;
    PLUS = 100; //OPERATOR
    MINUS = 101; //OPERATOR
    MULTIPLE = 102; //OPERATOR
    DIVIDE = 103; //OPERATOR
  }
  message Filter {
    enum OperType {
      KETI_DEFAULT = 0; //DEFAULT
      KETI_GE = 1;
      KETI_LE = 2;
      KETI_GT = 3;
      KETI_LT = 4;
      KETI_ET = 5;
      KETI_NE = 6;
      KETI_LIKE = 7;
      KETI_BETWEEN = 8;
      KETI_IN = 9;
      KETI_IS = 10;
      KETI_ISNOT = 11;
      KETI_NOT = 12;
      KETI_AND = 13;
      KETI_OR = 14;
    }
    message FilterValue{
      repeated ValueType type = 1;
      repeated string value = 2;
    }
    FilterValue LV = 1;
    OperType Operator = 2;
    FilterValue RV = 3;
    repeated FilterValue EXTRA = 4;
  }
  message Projection {
    enum SelectType{
      COLUMNNAME = 0; //DEFAULT
      SUM = 1;
      AVG = 2;
      COUNT = 3;
      COUNTSTAR = 4;
      TOP = 5;
      MIN = 6;
      MAX = 7;
      CASE = 8;
      WHEN = 9;
      THEN = 10;
      ELSE = 11;
      LIKE = 12;
    }
    SelectType select_type = 1;
    repeated string value = 2;
    repeated ValueType value_type = 3;    
  }
  message Order {
    enum OrderDirection{
      ASC = 0; //DEFAULT
      DESC = 1;
    }
    OrderDirection direction = 1;
    string column = 2;
  }
  int32 query_ID = 1;
  int32 work_ID = 2;
  repeated string table_name = 3;
  repeated string table_col = 4;
  repeated Filter table_filter = 5;
  repeated int32 table_offset = 6;
  repeated int32 table_offlen = 7;
  repeated int32 table_datatype = 8;
  string table_alias = 9;
  repeated string column_alias = 10;
  repeated Projection column_projection = 11;
  repeated string column_filtering = 12;
  repeated string group_by = 13;
  repeated Order order_by = 14;
}
message Request{
  int32 queryid = 1;
}
message Result {
  string value = 1;
}

